rfcs/2023-09-19-18604-metrics-data-model-refactor.md

```diff
@@ -0,0 +1,455 @@
+generally go about one of two ways: either breaking each bucket/statistic into an individual gauge
+(as this RFC proposes) or by converting it into another statistical data structure, such as
+`DDSketch`.
+
+By simplifying the data model, we can remove nearly all of that boilerplate handling code and simply
+let sinks handle each metric type in a more direct way. For counters and gauges, this is already
+taken care of as it stands today. For sinks handling either aggregated histograms, distributions, or
+sketches, they would now only need to deal with `Distribution`. Helper methods will be provided to
+convert `Distribution` into the highest fidelity form supported by the sink: a sink that supports
+sketches can ask for it to be converted to a sketch, while another sink that only supports
+histograms can ask for a histogram, since all distribution variants can be converted to either a
+histogram or sketch.
+### Aligns more closely with other metrics systems
+
+While the current data model is more of a union of possible metric types based on the metric systems
+we can ingest from, the observability world has been continually shifting over time to a simpler and
+more unified data model that simply deals with counters, gauges, and histograms. This can be seen in
+observability systems such as OpenTelemetry, or observed directly, such as by examining the allowed
+metric types for something like Datadog's ingestion APIs.
+
+While SDKs and helper code may provide higher-level primitives on top of these, such as rates and
+meters, we can simplify our underlying data model both for our own benefit, as well as the benefit
+of interoperability.
+
+## Drawbacks
+
+### Reassembling decomposed metrics for correct handling
+
+The main drawback would be the complexity required to handle the "recomposing" of individual metrics
+where an aggregated summary could otherwise be handled.
+
+For Prometheus-specific sinks, this would entail grouping the metrics together. This is trivial to
+do in the naive sense of, say, sorting them so that they're all adjacent in the sink output.
+However, the trickier part would be the logic to essentially "find" them and determine that they're
+the decomposed pieces of an aggregated histogram, rather than simply uncorrelated metrics. This is
+import for the ability to properly tag the metrics as being part of an aggregated histogram or
+summary.
+
+In general, there are other facets that would add risk to this recomposing logic, such as mutation
+of the metrics in a topology (tag governance/cost control modifications leading to missing
+individual metrics) or batching in time and/or space that leads to metrics losing their natural
+grouping.
+
+Some of these risks are greater than others: tag governance/cost control techniques in topologies
+are very common, but events being split out of their original batch is very rare. However, the risks
```

Comment:
I agree it's probably rare, but I worry about how difficult this would be to detect and whether we'd just need to live with the possibility that this recomposition would sometimes fail, seemingly at random.

To throw one idea out there, would it make sense to introduce a `MetricGroup` variant or similar that simply acts as an indivisible container for sets of related metrics like this?

Reply:
Well, I like the idea overall -- an indivisible container of sorts -- although it wouldn't be able to exist at the `MetricValue` level, or even the `Metric` level, because in this case, we need to be able to mark each metric with the different relevant tags as a means of differentiation.

I mean, technically speaking, we could probably have a `MetricValue` variant that holds a `Vec<Metric>` and contort ourselves to flatten that at the point of use... but in my mind, it feels like it would be very ugly.

Let me think a little more on this.

Reply:
Ah, good point. Would it be the series name that would differ? I imagine the tags would be the same, so it might actually be an advantage not to duplicate them. Maybe it could be a `MetricValue` if each entry included a `series_suffix` field or similar? Though that may be more complexity than it's worth.

Reply:
For the particular use case of Prometheus' aggregated summaries, you end up with, for a metric named `foo`:

- `foo_count` (no special tags)
- `foo_sum` (no special tags)
- multiple `foo` (`quantile` tag to differentiate which quantile the series represents)

So yeah, we'd need to indicate how to modify both the name _and_ the tags in the series, which is why it's a little tricky/gross overall.

Reply:
This change in handling of aggregated summaries, standing that metric model on its head, concerns me too.